ufl/algorithms/apply_derivatives.py

```diff
@@ -165,25 +190,65 @@ def independent_operator(self, o):
         """A zero with correct shape and indices for operators independent of diff. variable."""
         return Zero(o.ufl_shape + self._var_shape, o.ufl_free_indices, o.ufl_index_dimensions)
 
+    # --- Error checking for missing handlers and unexpected types
+
+    @singledispatchmethod
+    def process(self, o: Expr) -> Expr:
+        """Process ``o``.
+
+        Args:
+            o: `Expr` to be processed.
+
+        Returns:
+            Processed object.
+
+        """
+        raise AssertionError(f"UFL expression expected: got {o}")
+
+    @process.register(Expr)
+    def _(self, o: Expr) -> Expr:
+        """Raise error."""
+        raise ValueError(
+            f"Missing differentiation handler for type {o._ufl_class_.__name__}. "
+            "Have you added a new type?"
+        )
+
+    @process.register(Derivative)
+    def _(self, o: Expr) -> Expr:
```

Comment:
Wouldn't it be better to use 
```python
@process.register
    def _(self, o: Derivative) -> Expr:
```
here and in subsequent registrations, as https://docs.python.org/3/library/functools.html#functools.singledispatch states:
> To add overloaded implementations to the function, use the register() attribute of the generic function, which can be used as a decorator. For functions annotated with types, the decorator will infer the type of the first argument automatically:

I guess we would have to wait for lower version Python to be 3.11 to do this for unions (https://github.com/python/cpython/pull/30017). However, personally, I think this would be a cleaner solution. What do you think?

Reply:
I gave it a go with this as a separate commit, but I personally prefer:
```
    @process.register(Grad)
    @process.register(CellAvg)
    @process.register(FacetAvg)
    def _(self, o) -> Expr:
        ...
```
as type names are left-aligned and easier to read for me. I also do not quite see how it plays well when decorators are nested as in:
```
    @process.register
    @DAGTraverser.postorder
    @pending_operations_recording
    def external_operator(self, N: ExternalOperator, *dfs) -> Expr:
        ...
```
though `functools.wraps()` seems to be doing the job for us.

Please see the history for changes; I would like to revert this change back for now as Python3.10 test fails as you already pointed out.

Reply:
I think we should follow the Python convention here, which is the way Jørgen has suggested.

Reply:
Ok, it is good, but we indeed need Python >= 3.11, so we probably should not do this in this PR. I just dropped the relevant commit.

Reply:
We could make a note of this in https://github.com/FEniCS/ufl/issues/176 to remind us about this in October 2026

Reply:
Yes. I will drop a note once this PR is merged.